### ISSUE TITLE
Error "cannot evaluate a model without absence and presence data that are not NA"

### ISSUE THREAD

**Issue by PatWright** [jamiemkass/ENMeval#70]
Hello,

I'm struggling with an error message with ENMevaluate. I'm a bit confused because all the runs I've done previously have worked fine. It would be great to have an explanation of the error message and information on how to fix it.

Thanks,
Patrick

```
require(ENMeval)

>occ = occs_Hogs_#the presence data
>env =  LocalGrids1Final#the environmental grids
>bg.coords = bg_Hogs_#the background data
>bg.grp = bg_Hogs_.grp# the cross validation groups - presence
>occ.grp = occ.grp_Hogs_# the cross validation groups - background
 
 ##run with spatialy constrained data folds - LQ
> sc1_Hogs_LQ= ENMeval::ENMevaluate(occ, env, bg.coords,
+            occ.grp = occ.grp_Hogs_, bg.grp = bg_Hogs_.grp, 
+             RMvalues = seq(1, 2, 1),  #try regularisation multipliers of values 1 - 4, with intervals of 1
+ 		fc = c( "LQ"), #try these sets of feature types
+ 		categoricals = NULL,  method = "user", 
+ 		overlap = FALSE, aggregation.factor = c(2, 2), 
+ 		kfolds = NA, bin.output = TRUE, clamp = TRUE,
+ 		rasterPreds = TRUE, parallel = TRUE, numCores = 2, #numCores set to four to speed up processing
+ 		progbar = TRUE, updateProgress = TRUE, algorithm = 'maxent.jar')
Error in { : 
  task 1 failed - "cannot evaluate a model without absence and presence data that are not NA"
> traceback()
5: stop(simpleError(msg, call = expr))
4: e$fun(obj, substitute(ex), parent.frame(), e$data)
3: foreach(i = seq_len(length(args)), .packages = c("dismo", "raster", 
       "ENMeval", "rJava")) %dopar% {
       modelTune.maxentJar(pres, bg, env, nk, group.data, args[[i]], 
           userArgs, rasterPreds, clamp)
   }
2: tuning(occ, env, bg.coords, occ.grp, bg.grp, method, algorithm, 
       args, args.lab, categoricals, aggregation.factor, kfolds, 
       bin.output, clamp, alg, rasterPreds, parallel, numCores, 
       progbar, updateProgress, userArgs)
1: ENMeval::ENMevaluate(occ, env, bg.coords, occ.grp = occ.grp_Hogs_, 
       bg.grp = bg_Hogs_.grp, RMvalues = seq(1, 2, 1), fc = c("LQ"), 
       categoricals = NULL, method = "user", overlap = FALSE, aggregation.factor = c(2, 
           2), kfolds = NA, bin.output = TRUE, clamp = TRUE, rasterPreds = TRUE, 
       parallel = TRUE, numCores = 2, progbar = TRUE, updateProgress = TRUE, 
       algorithm = "maxent.jar")
```

**Comment by bobmuscarella** [comment]
Hi Patrick,

This error should mean that all occurrence or all background points in at least one of your data partitions has only NA values for at least one covariate.  If you want to send data and code that gives a reproducible example then maybe I can help track it down.  Otherwise, I recommend checking how many NA values you get when extracting env layer values to occ and bg.points per each occ.grp and bg.grp.

Good luck!
Bob

**Comment by PatWright** [comment]
Hi Bob,

I think I have figured out the issue. I had too many points that corresponded to NA values in my raster stack. I'm dealing with road buffers, so the precision of my points is not always great and they may sometimes be outside of the buffer in NAs.

Is there a set limit for the number or percentage of NAs with ENMevaluate?

Thanks,
Patrick

**Comment by ndimhypervol** [comment]
Patrick,

NA values for environmental variables (even for one of them) means the model function ignores these records. Thus, an occurrence dataset of 100 records with 75 NA values is equivalent to a dataset of 25 records.

**Comment by ndimhypervol** [comment]
Currently getting this as well for parallel processing, but cannot replicate it on test data, which is very strange. Will investigate.

**Comment by Stuart54** [comment]
I have the same problem trying to fit Maxent models. Other models such as raf, svm, net, etc work OK. I have tried to produce a minimal example:

library(flexsdm)
library(terra)
setwd("C:/temp/flexsdm")
rm(list=rm())

# Parameters ------------------
spname <- "Chrysotoxum arcuatum"
nbackg <- 5000
parts  <- 3

# Read observation data from Hoverfly Recording Scheme
# fields: species, x, y
# coordinates are centres of 1x1km squares of OSGB grid
obs <- read.csv("data/hrs_1km_2010-2023.csv")

# Load environmental layers
# -------------------------
# extent: GB
# coordinate system: OSGB in kms
# format: ESRI ASCII
# categorical layers are named "cat_xxx"
# (in this case, there is one categorical layer - dominant soil type)
layers <- list.files(path="layers",
                     pattern="*.asc",
                     full.names = TRUE)
layers <- rast(layers)

# Add environmental variables to observations
obs <- sdm_extract(data = obs,
                   x = "x",
                   y = "y",
                   env_layer = layers,
                   filter_na = TRUE)

# Presences for our species - and partition them
pres <- obs[obs$species == spname, 2:ncol(obs)]
pres$pa <- 1
pres <- part_random(data = pres,
                    pr_ab = "pa",
                    method = c(method= 'kfold', folds = parts))

# Background sample
# -----------------
# get the 1km squares from which data was received
bkgr  <- unique(obs[ ,2:ncol(obs)])
# take a sample of size: parts * nbackg
bkgr <- bkgr[sample(nrow(bkgr), nbackg*parts), ]
bkgr$pa <- 0
# partition it
bkgr <- part_random(data = bkgr,
                    pr_ab = "pa",
                    method = c(method= 'kfold', folds = parts))

# try Maxent model
mx <- fit_max(data = pres,
              response = "pa",
              predictors = names(layers)[-grep("^cat_", names(layers))],
              predictors_f = names(layers)[grep("^cat_", names(layers))],
              partition = ".part",
              background = bkgr,
              thr = "equal_sens_spec")

The result is: 

Formula used for model fitting:
~LCM_Acid_grassland + LCM_Arable_Horticulture + RainWetestQ + TMaxWarmestMon + TmeanAnn + I(LCM_Acid_grassland^2) + I(LCM_Arable_Horticulture^2) + I(RainWetestQ^2) + I(TMaxWarmestMon^2) + I(TmeanAnn^2) + hinge(LCM_Acid_grassland) + hinge(LCM_Arable_Horticulture) + hinge(RainWetestQ) + hinge(TMaxWarmestMon) + hinge(TmeanAnn) + LCM_Arable_Horticulture:LCM_Acid_grassland + RainWetestQ:LCM_Acid_grassland + TMaxWarmestMon:LCM_Acid_grassland + TmeanAnn:LCM_Acid_grassland + RainWetestQ:LCM_Arable_Horticulture + TMaxWarmestMon:LCM_Arable_Horticulture + TmeanAnn:LCM_Arable_Horticulture + TMaxWarmestMon:RainWetestQ + TmeanAnn:RainWetestQ + TmeanAnn:TMaxWarmestMon + categorical(cat_fao85fu) - 1

Replica number: 1/1
Partition number: 1/3
Partition number: 2/3
Partition number: 3/3
Error in sdm_eval(p = pred_test$pred[pred_test$pr_ab == 1], a = pred_test$pred[pred_test$pr_ab ==  : 
  cannot evaluate a model without absence and presence data that are not NA

There are no NAs in the data. Check this is the case:
> nrow(pres) 
[1] 607 
> sum(complete.cases(pres)) 
[1] 607 
> nrow(bkgr) 
[1] 15000 
> sum(complete.cases(bkgr)) 
[1] 15000

Stuart

**Comment by jamiemkass** [comment]
It's hard to diagnose because I'm not that familiar with flexsdm. Perhaps try posting this issue on the flexsdm repo?

**Comment by ShiminGu** [comment]
Hi Stuart,
I recently tried to use the 'flexsdm' package version 1.3.3 to run species distribution modeling. Maximum Entropy algorithm was applied and found the same reporting error as you met.
I wrote a text message to the authors of the package and I am waiting for a kind reply. 
By now, I notice the error may be solved by redefining the "response" variable name as "pr_ab".
Have a try. Good luck!

All the best,
Shimin
